### PR TITLE
[ONNX FE] Convert ONNX Attention op directly to SDPA instead of using function expansion

### PR DESCRIPTION
**Pull request by evkotov**
### Details:
- Add a direct frontend converter for the standard ONNX Attention op (opset 23) that produces ScaledDotProductAttention nodes, bypassing the buggy ONNX function expansion that fails on models like GPT-2 exported with dynamo at opset 23
- Support all attention modes (MHA, GQA with head repetition, MQA), 3D/4D inputs, KV cache, causal masking, float/bool masks, custom scale, and softcap
- Fall back to manual decomposition (matmul + softmax) when softcap or qk_matmul_output is requested, since SDPA does not support these directly
- Extract the duplicated get_dimensions helper from three attention-related converters into a shared utils/attention.hpp utility
- Add 11 test models (.prototxt) and corresponding test cases covering each feature

### Tickets:
 - 180686

### CODE REVIEW

File: src/frontends/onnx/frontend/src/ops_bridge.cpp [hunk 1]
```diff
@@ -166,8 +166,8 @@ OperatorSet OperatorsBridge::get_operator_set(const std::string& domain, int64_t
     for (const auto& op : dm->second) {
         const auto& it = find(version, op.second);
         if (it == std::end(op.second)) {
-            OPENVINO_THROW("Unsupported operator version: " + (domain.empty() ? "" : domain + ".") + op.first + ":" +
-                           std::to_string(version));
```
Comment (mvafin) [inline src/frontends/onnx/frontend/src/ops_bridge.cpp:170]: Why do we need this change?

Comment (evkotov) [reply src/frontends/onnx/frontend/src/ops_bridge.cpp:170]: removed

File: src/frontends/onnx/frontend/src/op/attention.cpp [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+
+    // Required inputs
+    auto Q = inputs[0];
+    auto K = inputs[1];
+    auto V = inputs[2];
+
+    // Optional inputs
+    bool has_attn_mask = num_inputs > 3 && !ov::op::util::is_null(inputs[3]);
+    bool has_past_key = num_inputs > 4 && !ov::op::util::is_null(inputs[4]);
+    bool has_past_value = num_inputs > 5 && !ov::op::util::is_null(inputs[5]);
```
Comment (mvafin) [inline src/frontends/onnx/frontend/src/op/attention.cpp]:
```suggestion
    bool has_attn_mask = common::is_input_valid(node, 3);
    bool has_past_key = common::is_input_valid(node, 4);
    bool has_past_value = common::is_input_valid(node, 5);
```

File: src/frontends/onnx/frontend/src/op/attention.cpp [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+    auto q_rank = Q.get_partial_shape().rank();
+    auto k_rank = K.get_partial_shape().rank();
+    auto v_rank = V.get_partial_shape().rank();
+    bool q_is_3d = false;
+    bool kv_is_3d = false;
+
+    if (q_rank.is_static()) {
+        q_is_3d = (q_rank.get_length() == 3);
+        CHECK_VALID_NODE(node,
+                         q_rank.get_length() == 3 || q_rank.get_length() == 4,
```
Comment (mvafin) [inline src/frontends/onnx/frontend/src/op/attention.cpp]:
```suggestion
                         q_is_3d || q_rank.get_length() == 4,
```

File: src/frontends/onnx/frontend/src/op/attention.cpp [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+                         "Q input rank must be 3 or 4, got: ",
+                         q_rank.get_length());
+    }
+    if (k_rank.is_static()) {
+        kv_is_3d = (k_rank.get_length() == 3);
+        CHECK_VALID_NODE(node,
+                         k_rank.get_length() == 3 || k_rank.get_length() == 4,
```
Comment (mvafin) [inline src/frontends/onnx/frontend/src/op/attention.cpp]:
```suggestion
                         kv_is_3d || k_rank.get_length() == 4,
```

File: src/frontends/onnx/frontend/src/op/attention.cpp [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+
+    CHECK_VALID_NODE(node, softcap >= 0.0f, "softcap must be non-negative, got: ", softcap);
+    CHECK_VALID_NODE(node,
+                     qk_matmul_output_mode >= 0 && qk_matmul_output_mode <= 3,
+                     "qk_matmul_output_mode must be 0, 1, 2, or 3, got: ",
+                     qk_matmul_output_mode);
+
+    // Determine number of requested outputs
+    size_t num_outputs = node.get_outputs_size();
+    bool needs_qk_output = num_outputs > 3;
```
[the remark this answers is not in the document]
Comment (mvafin) [reply src/frontends/onnx/frontend/src/op/attention.cpp]: Agree, the name needs to be checked

File: src/frontends/onnx/frontend/src/op/attention.cpp [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+    // as attn_mask instead, disabling SDPA's is_causal flag.
+    if (is_causal && has_past_key) {
+        auto causal_mask = detail::build_causal_mask(Q, K, true);
+        if (has_attn_mask) {
+            attn_mask = std::make_shared<v1::Add>(attn_mask, causal_mask);
+        } else {
+            attn_mask = causal_mask;
+        }
+        has_attn_mask = true;
+        is_causal = false;
```
[the remark this answers is not in the document]
Comment (mvafin) [reply src/frontends/onnx/frontend/src/op/attention.cpp]: @evkotov We have a test for boolean mask, right? I think this is a wrong comment
- Yes, we have a test for boolean mask (attention_mask_bool). Copilot's comment is wrong — the ONNX spec forbids combining is_causal=1 with attn_mask (line 96: assert attn_mask is None), so this code path can never produce Add(bool, float).

**Review by mvafin — COMMENTED** [review]
Looks good. Few unimportant fixes suggested. Please also address Copilot findings

File: src/frontends/onnx/frontend/src/op/attention.cpp [hunk 7]
```diff
@@ -0,0 +1,454 @@
...
+    auto input_shape = std::make_shared<v3::ShapeOf>(input);
+    auto batch = get_dimensions(input_shape, {0});
+    auto seq = get_dimensions(input_shape, {1});
+    auto num_heads_node = v0::Constant::create(ov::element::i64, ov::Shape{1}, {num_heads});
+    auto hidden = get_dimensions(input_shape, {2});
+    auto head_size = std::make_shared<v1::Divide>(hidden, num_heads_node);
+    // Direct reshape to (batch, num_heads, seq, head_size)
+    auto new_shape = std::make_shared<v0::Concat>(ov::NodeVector{batch, num_heads_node, seq, head_size}, 0);
+    return std::make_shared<v1::Reshape>(input, new_shape, false);
+}
```
[the remark this answers is not in the document]
Comment (evkotov) [reply src/frontends/onnx/frontend/src/op/attention.cpp:57]: The asymmetry between reshape_3d_to_4d (direct reshape) and reshape_4d_to_3d (transpose + reshape) is intentional and matches the ONNX reference implementation exactly. See op_attention.py: forward at lines 57-58 uses direct reshape, backward at lines 174-175 uses np.transpose(output, (0, 2, 1, 3)) followed by np.reshape. These are not meant to be mathematical inverses — the attention computation between them changes the data semantics.

File: src/frontends/onnx/frontend/src/op/attention.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (evkotov) [reply src/frontends/onnx/frontend/src/op/attention.cpp]: @mvafin Correct, this is a wrong comment. The ONNX spec explicitly forbids combining is_causal=1 with attn_mask (see op_attention.py line 96: assert attn_mask is None when is_causal == 1). So this code path with Add(bool, float) can never be reached. We do have a test for boolean masks (attention_mask_bool) and it passes. Additionally, this code is now outdated — the condition was changed to handle all is_causal cases uniformly.

File: src/frontends/onnx/frontend/src/op/attention.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (evkotov) [reply src/frontends/onnx/frontend/src/op/attention.cpp]: fixed
